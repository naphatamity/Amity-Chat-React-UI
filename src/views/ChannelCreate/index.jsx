import { useState, useMemo } from 'react';
import { Link, useHistory } from 'react-router-dom';
import rug from 'random-username-generator';
import { ChannelRepository, ChannelType } from '@amityco/js-sdk';
import ChannelHeader from '../../components/ChannelHeader';
import UsersListRow from '../../components/UsersListRow';
import SelectedUserItem from '../../components/SelectedUserItem';

import { useConfig } from '../../components/Config';
import {
  Wrapper,
  CreateChannelWrapper,
  SearchInput,
  SelectedUsersList,
  LetterDivider,
  UsersList,
  Title,
  TitleBar,
  CreateButton,
  ScrollContainer,
  StretchedList,
} from './styles';

const channelRepo = new ChannelRepository();

const users = [
  { userId: 'bryan', displayName: 'bryan' },
  { userId: 'floyd', displayName: 'floyd' },
  { userId: 'jess', displayName: 'jess' },
  { userId: 'lena', displayName: 'lena' },
  { userId: 'liz', displayName: 'liz' },
  { userId: 'marga', displayName: 'marga' },
  { userId: 'mimi', displayName: 'mimi' },
  { userId: 'mssmith', displayName: 'mssmith' },
  { userId: 'nick', displayName: 'nick' },
  { userId: 'theo', displayName: 'theo' },
];

function groupByLetter(array) {
  return array.reduce((acc, element) => {
    const group = element.displayName ? element.displayName.trim()[0] : element.userId.trim()[0];
    if (!acc[group]) acc[group] = { group, users: [element] };
    else acc[group].users.push(element);
    return acc;
  }, {});
}

// ==========================
function ChannelCreate() {
  const { client } = useConfig();
  const history = useHistory();
  const [selected, setSelected] = useState([]);
  const [search, setSearch] = useState('');
  const [newChannelId, setNewChannelId] = useState('');
  const usersByLetters = useMemo(() => {
    const filteredUsers = users.filter(item => !search || item.displayName === search);
    return groupByLetter(filteredUsers);
  }, [search]);

  const isSelected = userId => {
    if (selected.indexOf(userId) < 0) {
      return false;
    }
    return true;
  };

  const onSelect = userId => {
    if (!isSelected(userId)) {
      const newArray = [...selected, userId];
      setSelected(newArray);
    }
  };

  const onUnselect = userId => {
    if (!isSelected(userId)) {
      return null;
    }

    setSelected(selected.filter(item => item !== userId));
  };

  const create = () => {
    const userIds = [client.currentUser.model.userId, ...selected];
    const generatedName = rug.generate();
    const createChat = channelRepo.createChannel({
      channelId: newChannelId || generatedName,
      type: ChannelType.Standard,
      userIds,
    });
    createChat.once('dataUpdated', () => {
      history.push('/channels');
    });
  };

  return (
    <Wrapper>
      <ChannelHeader>
        <Link to="/channels" style={{ textDecoration: 'none', color: '#17181C' }}>
          Cancel
        </Link>

        <TitleBar>
          <Title>{`${selected.length} Selected`}</Title>
        </TitleBar>
        <CreateButton onClick={() => create()}>Create</CreateButton>
      </ChannelHeader>
      <CreateChannelWrapper>
        <SearchInput
          style={{ marginBottom: 16 }}
          placeholder="Group name"
          onChange={e => setNewChannelId(e.target.value)}
        />
        <SearchInput placeholder="Search" onChange={e => setSearch(e.target.value)} />
        <SelectedUsersList>
          <ScrollContainer>
            <StretchedList>
              {selected.map(userId => (
                <SelectedUserItem key={userId} userId={userId} onUnselect={onUnselect} />
              ))}
            </StretchedList>
          </ScrollContainer>
        </SelectedUsersList>
        {Object.keys(usersByLetters).map(letter => (
          <div key={letter}>
            <LetterDivider>{letter}</LetterDivider>
            <UsersList>
              {usersByLetters[letter].users.map(({ userId }) => (
                <UsersListRow
                  key={userId}
                  userId={userId}
                  isSelected={isSelected(userId)}
                  onClick={() => (isSelected(userId) ? onUnselect(userId) : onSelect(userId))}
                />
              ))}
            </UsersList>
          </div>
        ))}
      </CreateChannelWrapper>
    </Wrapper>
  );
}

export default ChannelCreate;
