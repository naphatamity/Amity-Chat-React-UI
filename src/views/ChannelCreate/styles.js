import styled from 'styled-components';

export const SearchInput = styled.input`
  border: 1px solid #ebecef;
  border-radius: 4px;
  padding: 10px 12px;
  flex-grow: 1;
  background: #ebecef;
  width: 100%;
  box-sizing: border-box;
  margin-bottom: 16px;
`;

export const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  flex-grow: 1;
`;

export const CreateChannelWrapper = styled.div`
  padding: 16px;
  position: relative;
`;

export const ScrollContainer = styled.div`
  width: 100%;
  overflow: auto;
  position: relative;
`;

export const StretchedList = styled.div`
  display: flex;
  align-items: flex-start;
  justify-content: flex-start;

  & > * {
    flex: 0 0 auto;
  }
`;

export const SelectedUsersList = styled.div`
  list-style: none;
  padding: 0;
  margin: 0;
  margin-bottom: 8px;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  min-height: 74px;
  position: relative;
  // &:before {
  //   content: '';
  //   position: absolute;
  //   bottom: 0;
  //   right: 0;
  //   background: linear-gradient(to right, transparent 40%, white);
  //   height: 100%;
  //   width: 15%;
  //   z-index: 999;
  // }
`;

export const SelectedUserItem = styled.div`
  padding: 8px;
  position: relative;
  overflow: hidden;
  text-overflow: ellipsis;
  width: 64px;
  height: 62px;
  font-size: 13px;
  margin-right: 16px;

}
`;

export const LetterDivider = styled.div`
  text-transform: uppercase;
  background: #ebecef;
  font-style: normal;
  font-weight: 600;
  font-size: 15px;
  line-height: 20px;
  letter-spacing: -0.24px;
  color: #a5a9b5;
  width: 100%;
  margin-left: -16px;
  padding: 6px 16px;
  margin-bottom: 12px;
`;

export const UsersList = styled.ul`
  list-style: none;
  padding: 0;
  margin: 0;
`;
export const UsersListRow = styled.li`
  display: flex;
  flex-direction: row;
  align-items: center;
  margin-bottom: 16px;

  &:last-of-type {
    margin-bottom: 12px;
  }
`;

export const UsersListName = styled.div`
  margin-left: 8px;
`;

export const TitleBar = styled.div`
  display: flex;
  flex-flow: row;
  flex-grow: 1;
  align-items: center;
  justify-content: center;
`;

export const Title = styled.h2`
  margin: 0;
  margin-left: 8px;
  font-style: normal;
  font-weight: 600;
  font-size: 17px;
  line-height: 22px;
  letter-spacing: -0.41px;
`;

export const UnselectButton = styled.button`
  background: #292b32;
  opacity: 0.5;
  height: 16px;
  width: 16px;
  border-radius: 50%;
  color: white;
  border: none;
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  top: 4px;
  position: absolute;
  right: 4px;
`;

export const CreateButton = styled.button`
  background: transparent;
  border: none;
  font-size: 15px;
  color: #1054de;
`;
