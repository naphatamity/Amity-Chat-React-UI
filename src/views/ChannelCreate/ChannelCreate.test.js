import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';

import ChannelCreate from '.';

describe('<ChannelCreate /> spec', () => {
  it('renders the component', () => {
    const container = render(
      <Router>
        <ChannelCreate />
      </Router>,
    );
    expect(container).toMatchSnapshot();
  });
});
