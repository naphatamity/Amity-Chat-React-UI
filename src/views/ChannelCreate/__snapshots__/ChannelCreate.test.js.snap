// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChannelCreate /> spec renders the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-jrQzUz jHgdLV"
      >
        <div
          class="sc-bdvvaa fiTrLn"
        >
          <a
            href="/channels"
            style="text-decoration: none; color: rgb(23, 24, 28);"
          >
            Cancel
          </a>
          <div
            class="sc-bkkfTU dzqiXj"
          >
            <h2
              class="sc-ieebsP kWkvkc"
            >
              0 Selected
            </h2>
          </div>
          <button
            class="sc-hGPAah fFSkrZ"
          >
            Create
          </button>
        </div>
        <div
          class="sc-kDThTU jCuAAp"
        >
          <input
            class="sc-pVTma htPafV"
            placeholder="Group name"
            style="margin-bottom: 16px;"
          />
          <input
            class="sc-pVTma htPafV"
            placeholder="Search"
          />
          <div
            class="sc-egiSv gsIIzT"
          >
            <div
              class="sc-iqsfdx hiUiZe"
            >
              <div
                class="sc-crHlIS bYQTIM"
              />
            </div>
          </div>
          <div>
            <div
              class="sc-ksdxAp fHHPPY"
            >
              b
            </div>
            <ul
              class="sc-hBURRC iiJnpa"
            >
              <li
                class="sc-gsDJrp gZOWSV"
              >
                Loading
              </li>
            </ul>
          </div>
          <div>
            <div
              class="sc-ksdxAp fHHPPY"
            >
              f
            </div>
            <ul
              class="sc-hBURRC iiJnpa"
            >
              <li
                class="sc-gsDJrp gZOWSV"
              >
                Loading
              </li>
            </ul>
          </div>
          <div>
            <div
              class="sc-ksdxAp fHHPPY"
            >
              j
            </div>
            <ul
              class="sc-hBURRC iiJnpa"
            >
              <li
                class="sc-gsDJrp gZOWSV"
              >
                Loading
              </li>
            </ul>
          </div>
          <div>
            <div
              class="sc-ksdxAp fHHPPY"
            >
              l
            </div>
            <ul
              class="sc-hBURRC iiJnpa"
            >
              <li
                class="sc-gsDJrp gZOWSV"
              >
                Loading
              </li>
              <li
                class="sc-gsDJrp gZOWSV"
              >
                Loading
              </li>
            </ul>
          </div>
          <div>
            <div
              class="sc-ksdxAp fHHPPY"
            >
              m
            </div>
            <ul
              class="sc-hBURRC iiJnpa"
            >
              <li
                class="sc-gsDJrp gZOWSV"
              >
                Loading
              </li>
              <li
                class="sc-gsDJrp gZOWSV"
              >
                Loading
              </li>
              <li
                class="sc-gsDJrp gZOWSV"
              >
                Loading
              </li>
            </ul>
          </div>
          <div>
            <div
              class="sc-ksdxAp fHHPPY"
            >
              n
            </div>
            <ul
              class="sc-hBURRC iiJnpa"
            >
              <li
                class="sc-gsDJrp gZOWSV"
              >
                Loading
              </li>
            </ul>
          </div>
          <div>
            <div
              class="sc-ksdxAp fHHPPY"
            >
              t
            </div>
            <ul
              class="sc-hBURRC iiJnpa"
            >
              <li
                class="sc-gsDJrp gZOWSV"
              >
                Loading
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-jrQzUz jHgdLV"
    >
      <div
        class="sc-bdvvaa fiTrLn"
      >
        <a
          href="/channels"
          style="text-decoration: none; color: rgb(23, 24, 28);"
        >
          Cancel
        </a>
        <div
          class="sc-bkkfTU dzqiXj"
        >
          <h2
            class="sc-ieebsP kWkvkc"
          >
            0 Selected
          </h2>
        </div>
        <button
          class="sc-hGPAah fFSkrZ"
        >
          Create
        </button>
      </div>
      <div
        class="sc-kDThTU jCuAAp"
      >
        <input
          class="sc-pVTma htPafV"
          placeholder="Group name"
          style="margin-bottom: 16px;"
        />
        <input
          class="sc-pVTma htPafV"
          placeholder="Search"
        />
        <div
          class="sc-egiSv gsIIzT"
        >
          <div
            class="sc-iqsfdx hiUiZe"
          >
            <div
              class="sc-crHlIS bYQTIM"
            />
          </div>
        </div>
        <div>
          <div
            class="sc-ksdxAp fHHPPY"
          >
            b
          </div>
          <ul
            class="sc-hBURRC iiJnpa"
          >
            <li
              class="sc-gsDJrp gZOWSV"
            >
              Loading
            </li>
          </ul>
        </div>
        <div>
          <div
            class="sc-ksdxAp fHHPPY"
          >
            f
          </div>
          <ul
            class="sc-hBURRC iiJnpa"
          >
            <li
              class="sc-gsDJrp gZOWSV"
            >
              Loading
            </li>
          </ul>
        </div>
        <div>
          <div
            class="sc-ksdxAp fHHPPY"
          >
            j
          </div>
          <ul
            class="sc-hBURRC iiJnpa"
          >
            <li
              class="sc-gsDJrp gZOWSV"
            >
              Loading
            </li>
          </ul>
        </div>
        <div>
          <div
            class="sc-ksdxAp fHHPPY"
          >
            l
          </div>
          <ul
            class="sc-hBURRC iiJnpa"
          >
            <li
              class="sc-gsDJrp gZOWSV"
            >
              Loading
            </li>
            <li
              class="sc-gsDJrp gZOWSV"
            >
              Loading
            </li>
          </ul>
        </div>
        <div>
          <div
            class="sc-ksdxAp fHHPPY"
          >
            m
          </div>
          <ul
            class="sc-hBURRC iiJnpa"
          >
            <li
              class="sc-gsDJrp gZOWSV"
            >
              Loading
            </li>
            <li
              class="sc-gsDJrp gZOWSV"
            >
              Loading
            </li>
            <li
              class="sc-gsDJrp gZOWSV"
            >
              Loading
            </li>
          </ul>
        </div>
        <div>
          <div
            class="sc-ksdxAp fHHPPY"
          >
            n
          </div>
          <ul
            class="sc-hBURRC iiJnpa"
          >
            <li
              class="sc-gsDJrp gZOWSV"
            >
              Loading
            </li>
          </ul>
        </div>
        <div>
          <div
            class="sc-ksdxAp fHHPPY"
          >
            t
          </div>
          <ul
            class="sc-hBURRC iiJnpa"
          >
            <li
              class="sc-gsDJrp gZOWSV"
            >
              Loading
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
